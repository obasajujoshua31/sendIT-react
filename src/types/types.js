export const signUpUser = "SIGN_UP_USER";
export const signInUser = "SIGN_IN_USER";
export const createOrder = "CREATE_ORDER";
export const createOrderSuccess = "CREATE_ORDER_SUCCESS";
export const createOrderFailure = "CREATE_ORDER_FAILURE";
export const viewOrders = "VIEW_ORDERS";
export const viewOrder = "VIEW_ORDER";
export const updateParcel = "UPDATE_PARCELS";
export const updateLocation = "UPDATE_PARCEL_LOCATION";
export const cancelOrder = "CANCEL_ORDER";
export const updateParcelStatus = "UPDATE_PARCEL_STATUS";
export const signInSuccess = "SIGN_IN_SUCCESS";
export const signInFailure = "SIGN_IN_FAILURE";
export const signUpSuccess = "SIGN_UP_SUCCESS";
export const signUpFailure = "SIGN_UP_FAILURE";
export const isLoggedOut = "LOGGED_OUT";
export const isAdmin = "IS_ADMIN";
