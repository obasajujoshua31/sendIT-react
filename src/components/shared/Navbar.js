import React from "react";
import { Link, withRouter } from "react-router-dom";
import { connect } from "react-redux";
import logOutUser from "../../actions/user/signout";
import { isLoggedOut, signInSuccess, signUpSuccess } from "../../types/types";

const GuestLink = (
  <React.Fragment>
    <li className="nav-item">
      <Link className="nav-link" to="/">
        Home <span className="sr-only">(current)</span>
      </Link>
    </li>
    <li className="nav-item">
      <Link className="nav-link" to="/about">
        About Us
      </Link>
    </li>
    <li className="nav-item">
      <Link className="nav-link" to="/signup">
        Create Account
      </Link>
    </li>{" "}
    <li className="nav-item">
      <Link className="nav-link" to="/login">
        Sign in
      </Link>
    </li>
  </React.Fragment>
);

export const Navbar = ({ logout, singInStatus, history, role }) => {
  const singInUserLink = (
    <React.Fragment>
      {role ? (
        <li className="nav-item">
          <Link to="/admin-dashboard" className="nav-link">
            Parcel Summary
          </Link>
        </li>
      ) : (
        <li className="nav-item">
          <Link to="/dashboard" className="nav-link">
            Parcel Summary
          </Link>
        </li>
      )}
      {!role && (
        <li className="nav-item">
          <Link to="/create-order" className="nav-link">
            Send Parcels
          </Link>
        </li>
      )}
      {role ? (
        <li className="nav-item">
          <Link to="/admin-view-orders" className="nav-link">
            View Orders
          </Link>
        </li>
      ) : (
        <li className="nav-item">
          <Link to="/view-orders" className="nav-link">
            View Orders
          </Link>
        </li>
      )}

      <li className="nav-item">
        <a
          className="nav-link"
          onClick={() => {
            logout();
            history.push("/");
          }}
        >
          Logout
        </a>
      </li>
    </React.Fragment>
  );

  return (
    <nav
      className="navbar navbar-expand-lg navbar-light"
      style={{ backgroundColor: "#806B75" }}
    >
      <button
        className="navbar-toggler"
        type="button"
        data-toggle="collapse"
        data-target="#navbarNavDropdown"
        aria-controls="navbarNavDropdown"
        aria-expanded="false"
        aria-label="Toggle navigation"
      >
        <span className="navbar-toggler-icon" />
      </button>
      <div className="collapse navbar-collapse" id="navbarNavDropdown">
        <ul className="navbar-nav">
          {singInStatus === signInSuccess || singInStatus === signUpSuccess
            ? singInUserLink
            : GuestLink}
        </ul>
      </div>
    </nav>
  );
};
export const mapStateToProps = ({ users }) => ({
  singInStatus: users.singInStatus,
  role: users.role
});

export const mapDispatchToProps = dispatch => ({
  logout: () => dispatch(logOutUser(isLoggedOut))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(Navbar));
