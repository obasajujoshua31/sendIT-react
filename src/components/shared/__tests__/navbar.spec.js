import React from "react";
import { shallow, mount } from "enzyme";
import sinon from "sinon";
import { Navbar, mapDispatchToProps, mapStateToProps } from "../Navbar";
import { signInSuccess } from "../../../types/types";

const props = {
  logout: jest.fn(),
  singInstatus: "",
  history: {
    push: jest.fn()
  },
  role: ""
};

const mockStore = {
  users: {
    singInStatus: "Success",
    role: false
  }
};

describe("<Navbar>", () => {
  const wrapper = shallow(<Navbar {...props} />);
  it("should render a navbar component", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("should contain 1 <div>", () => {
    expect(wrapper.find("div").length).toEqual(1);
  });
  it("should contain 4 <li> tag", () => {
    expect(wrapper.find("li").length).toEqual(4);
  });
  it("should contain 1 <nav> tag", () => {
    expect(wrapper.find("nav").length).toEqual(1);
  });
  it("should contain 4 <Link> tags", () => {
    expect(wrapper.find("Link").length).toEqual(4);
  });
  it("should call dispatch action", () => {
    const dispatchSpy = sinon.spy();
    // console.log(mapDispatchToProps)
    const dispatchResult = mapDispatchToProps(dispatchSpy);
    // console.log(dispatchResult);
    dispatchResult.logout();
    expect(dispatchSpy.callCount).toEqual(1);
  });
  it("should map state to Props", () => {
    const mapStateResult = mapStateToProps(mockStore);
    expect(mapStateResult.singInStatus).toEqual("Success");
  });

  it("should call logout", () => {
    const alteredProps = { ...props, singInStatus: signInSuccess };
    const navWrapper = shallow(<Navbar {...alteredProps} />);

    navWrapper.find("a").simulate("click");
    // expect(navWrapper.instance().props.history.push).toHaveBeenCalled();
  });
  it("should render 3 li tags if it is Admin", () => {
    const alteredProps = { ...props, singInStatus: signInSuccess, role: true };
    const navWrapper = shallow(<Navbar {...alteredProps} />);

    expect(navWrapper.find("li").length).toEqual(3);
  });
});
