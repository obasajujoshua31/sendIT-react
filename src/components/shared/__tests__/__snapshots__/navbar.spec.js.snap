// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar> should render a navbar component 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-light"
  style={
    Object {
      "backgroundColor": "#806B75",
    }
  }
>
  <button
    aria-controls="navbarNavDropdown"
    aria-expanded="false"
    aria-label="Toggle navigation"
    className="navbar-toggler"
    data-target="#navbarNavDropdown"
    data-toggle="collapse"
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    className="collapse navbar-collapse"
    id="navbarNavDropdown"
  >
    <ul
      className="navbar-nav"
    >
      <li
        className="nav-item"
      >
        <Link
          className="nav-link"
          replace={false}
          to="/"
        >
          Home 
          <span
            className="sr-only"
          >
            (current)
          </span>
        </Link>
      </li>
      <li
        className="nav-item"
      >
        <Link
          className="nav-link"
          replace={false}
          to="/about"
        >
          About Us
        </Link>
      </li>
      <li
        className="nav-item"
      >
        <Link
          className="nav-link"
          replace={false}
          to="/signup"
        >
          Create Account
        </Link>
      </li>
       
      <li
        className="nav-item"
      >
        <Link
          className="nav-link"
          replace={false}
          to="/login"
        >
          Sign in
        </Link>
      </li>
    </ul>
  </div>
</nav>
`;
